### ISSUE TITLE
BFGS_minimize does not terminate

### ISSUE THREAD

**Issue by Hoeze** [tensorflow/probability#151]
I currently tried to implement your BFGS optimizer in a simple negative binomial GLM with 2000 observations and 10 parameters:
`loss = - mean(NB(X, mu=exp(design_loc*a), phi=exp(design_scale*b))`

The SciPy L-BFGS solves this GLM with seven steps in fractions of a second.
Running `bfgs_minimize`, however, does not terminate.

Are there any differences to the SciPy L-BFGS implementation?

---
Simplified (pseudo-)code:
```python
X = <sample data>
design_loc = <design matrix of mu>
design_scale = <design matrix of phi>
a = tf.Variable(...)
b = tf.Variable(...)

param_vec = tf.concat([a, b], axis=0, name="param_vec")
a_split, b_split = tf.split(param_vec, tf.TensorShape([a.shape[0], b.shape[0]]))

#loss = loss_fn(X, a_split, b_split)
#hess = tf.hessians(loss, param_vec)[0]

def value_and_grad_fn(param_vec):
    a_split, b_split = tf.split(param_vec, tf.TensorShape([a.shape[0], b.shape[0]]))
    loss = loss_fn(X, a_split, b_split)
    return loss, tf.gradients(loss, param_vec)

bfgs_res = bfgs_minimize(value_and_grad_fn, param_vec)
sess = tf.Session()
sess.run(tf.global_variables_initializer())
sess.run(bfgs_res)
```
I've added `tf.Print(retval, [converged, failed, num_iterations], "converged | failed | num_iterations:\t")` to `_cond` in `bfgs_minimize` in order to get some information about the loop evaluations. Result:
```
converged | failed | num_iterations:	[0][0][0]
converged | failed | num_iterations:	[0][0][1]
converged | failed | num_iterations:	[0][0][2]
converged | failed | num_iterations:	[0][0][3]
```

The first three to four steps always happen in fractions of a second.
Then the session gets stuck in one loop iteration, running all CPU cores at 100%...

**Comment by saxena-ashish-g** [comment]
The scipy bfgs (and l-bfgs) optimizer uses a different line search algorithm (More Thuente) than the one that is implemented in TFP (Hager Zhang) so exact match between the two is not expected. However, the behaviour you encountered is definitely a bug. 

I have located the source of this problem and will be committing a fix in the next few days. Will update the thread once that is checked in and you can give your example another spin.

Thanks!

**Comment by Hoeze** [comment]
Thank you for looking into this :)

**Comment by saxena-ashish-g** [comment]
Can you please try your example again after incorporating this [commit](https://github.com/tensorflow/probability/commit/8af0224d4a9aa8f251b3cb22d997827cb51040a3)?

If it is fixed, please close the issue.

**Comment by Hoeze** [comment]
Sadly, this problem still occurs.

Should I try to write a non-working example?

**Comment by saxena-ashish-g** [comment]
Yes please. If you could flesh out a complete minimal example demonstrating the issue, it would be much easier to get to the root of it. Thanks!

**Comment by Hoeze** [comment]
Did you install the `netcdf4` package? Sorry, I forgot to mention that:
http://xarray.pydata.org/en/stable/io.html#netcdf
![grafik](https://user-images.githubusercontent.com/1200058/45310864-be8c0700-b527-11e8-8812-50f970915907.png)

I'll update the examples with BFGS. 
Just tell me if installing `pip3 install netcdf4` does not work as well, then I can try something else.

**Comment by Hoeze** [comment]
OK, the update is online.

- Using float32, scipy BFGS actually complains about missing numeric precision
- Using float64, the scipy BFGS is working. 
Also, the previously non-terminating example now works with float64 (`bfgs_test_float64_working.py`).
However, when initializing with something else than the MLE, the Tensorflow BFGS still fails (`bfgs_test_float64_failing.py`, `bfgs_test_float64_failing_with_hessian.py`).

**Comment by saxena-ashish-g** [comment]
Thanks for your patience. I have [pushed a significant overhaul](https://github.com/tensorflow/probability/commit/0b107bbc84d496d6428cd37e4090781b951751f6) which (hopefully) fixes the problems you were seeing.

Note that in your example where  you supply the inverse hessian there are a couple of issues:

1. You are supplying the initial Hessian and not the inverse.
2. The initial Hessian is actually not positive definite which violates one of the important assumptions behind the BFGS algorithm.

To help guard against (2), there is now an explicit check that the initial Hessian if supplied, must be positive definite and symmetric.

Please give it another spin and let me know if  you see more issues.

**Comment by Hoeze** [comment]
Wow, thanks for putting so much work on this!
I'll test it as soon as I can :+1:

**Comment by saxena-ashish-g** [comment]
I am going to mark this issue as closed. Please feel free to reopen it if you find more problems.

**Comment by Hoeze** [comment]
OK, I'm currently finishing my thesis so I had no time to try it out yet
